File: app/src/prod/java/torille.fi.lurkforreddit/RedditRepositoryModule.java
package torille.fi.lurkforreddit;

import dagger.Binds;
import dagger.Module;
import torille.fi.lurkforreddit.data.RedditDataSource;
import torille.fi.lurkforreddit.data.Remote;
import torille.fi.lurkforreddit.data.remote.RedditRemoteDataSource;
import torille.fi.lurkforreddit.di.scope.RedditScope;

/**
 * Created by eva on 23.4.2017.
 */
@Module
abstract public class RedditRepositoryModule {

    @RedditScope
    @Binds
    @Remote
    abstract RedditDataSource provideRemoteDataSource(RedditRemoteDataSource dataSource);

}

File: app/src/prod/java/torille.fi.lurkforreddit/RedditRepositoryModule.kt
package torille.fi.lurkforreddit

import dagger.Module
import dagger.Provides
import torille.fi.lurkforreddit.data.RedditDataSource
import torille.fi.lurkforreddit.data.RedditService
import torille.fi.lurkforreddit.data.SubredditDao
import torille.fi.lurkforreddit.data.local.RedditLocalDataSource
import torille.fi.lurkforreddit.data.remote.RedditRemoteDataSource
import torille.fi.lurkforreddit.di.scope.Local
import torille.fi.lurkforreddit.di.scope.Remote
import torille.fi.lurkforreddit.utils.CommentsStreamingParser
import torille.fi.lurkforreddit.utils.Store
import javax.inject.Singleton

/**
 * This is used by Dagger to inject the required arguments
 */
@Module
class RedditRepositoryModule {

    @Singleton
    @Provides
    @Remote
    internal fun provideRemoteDataSource(
        commentsStreamingParser: CommentsStreamingParser,
        redditApi: RedditService.Reddit,
        settingsStore: Store
    ): RedditDataSource {
        return RedditRemoteDataSource(redditApi, settingsStore, commentsStreamingParser)
    }

    @Singleton
    @Provides
    @Local
    internal fun provideLocalDataSource(subredditDao: SubredditDao): RedditDataSource {
        return RedditLocalDataSource(subredditDao)
    }

}
